### ISSUE TITLE
[Bug] Maximum call stack size exceeded

### ISSUE THREAD

**Issue by MarcMogdanz** [Kong/insomnia#1367]
- Insomnia Version: [insomnia-importers 2.0.14](https://www.npmjs.com/package/insomnia-importers/v/2.0.14)
- Node Version: 11.6.0

## Details

When running the code below with a file `specification.yml` which is a swagger 2.0 documentation I retrieve the error below. The original file size was ~110kb but I also reduced the size to around ~15kb but it still didn't work. Someone else encountered this problem also but he didn't create an issue, the comment can be found [here](https://github.com/getinsomnia/insomnia/issues/281#issuecomment-452886016).

```javascript
const fs = require("fs");
const importers = require("insomnia-importers");

fs.readFile("specification.yml", async (err, data) => {
  if (err) throw err;

  try {
    const output = await importers.convert(data.toString());
    console.log(JSON.stringify(output.data, null, 2));
  } catch (error) {
    console.log("error", error);
  }
});
```

```
error RangeError: Maximum call stack size exceeded
    at Array.forEach (<anonymous>)
    at object (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:300:31)
    at generateParameterExample (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:332:12)
    at Object.keys.forEach.propertyName (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:301:33)
    at Array.forEach (<anonymous>)
    at object (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:300:31)
    at generateParameterExample (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:332:12)
    at Object.keys.forEach.propertyName (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:301:33)
    at Array.forEach (<anonymous>)
    at object (/PATH/node_modules/insomnia-importers/src/importers/swagger2.js:300:31)
```

**Comment by gschier** [comment]
Can you provide a sample import file that can be used to reproduce this?

**Comment by ac1dr3d** [comment]
> Can you provide a sample import file that can be used to reproduce this?

Any big config file will do.

**Comment by gschier** [comment]
Without a file to use to reproduce the issue, there's no way I can help fix it. Please provide a sample file if this is still an issue.

**Comment by tobylo** [comment]
Just stumbled upon this myself, e.g. https://github.com/MicrosoftDocs/vsts-rest-api-specs/blob/master/specification/build/5.0/build.json will fail with "Maximum call stack size exceeded".

**Comment by shasureshkumar** [comment]
Even I am facing same kind of issue when importing huge data, as I am trying to deploy my application's swagger file I couldn't share here.
<img width="782" alt="Screenshot 2020-04-29 at 9 13 40 PM" src="https://user-images.githubusercontent.com/11661366/81699588-3214b300-9485-11ea-91e4-bb1a3d021a43.png">

**Comment by gillescoolen** [comment]
@gschier could you take a look at the example file @tobylo provided? Thanks!

**Comment by soracoder** [comment]
Still an issue. How is this going? has anyone take a look at this since 2019?

**Comment by ac1dr3d** [comment]
> Still an issue. How is this going? has anyone take a look at this since 2019?

nope. no one cares.
tho its big drawback... :/ and just needs small fix.
the only way i suppose to make a pull request.

**Comment by TheAifam5** [comment]
I have same issue for OpenAPI 3 and Swagger 2.0 documents :/

**Comment by tilakdewangan** [comment]
I am also facing the same issue. My finding is - when you have a relation between 2 classes and they have each other as a property then Swagger UI stuck in a loop.

**Comment by jcsena** [comment]
Same problem trying  to upload `swagger.json`

![image](https://user-images.githubusercontent.com/4293508/94082373-6f3f6900-fdc6-11ea-8283-c50b0f5e75bf.png)

**Comment by TheAifam5** [comment]
The problem occurs while generating examples for an object, calls again the function and the loop never closes

**Comment by Xetera** [comment]
> Just stumbled upon this myself, e.g. https://github.com/MicrosoftDocs/vsts-rest-api-specs/blob/master/specification/build/5.0/build.json will fail with "Maximum call stack size exceeded".

I'm not sure if this was fixed but as of the latest update 2021.1, I can import this file without getting any errors

**Comment by DonaldFoss** [comment]
I'm getting the error right now with 2021.1.1. I have 7 files my client provided. 3 of them generate this error. I cannot post the files as examples due to NDA. 

I believe the file is valid because VS Code recognizes the file, and it's plugins show the generated documentation. I can also open the file using the web interface for the latest Swagger Editor using a docker image.

In Insomnia, if I create a private sub-environment, the error still appears. If I go to Dashboard->Create->Import from File, the error still occurs. Using a JSON file that Swagger Editor exported from the YAML file, Insomnia reports the same error when I try to import it. This tells me that the issue is outside of reading and parsing YAML and JSON files.

To summarize, the file appears to be valid, so Insomnia's error stating it is an invalid file is generic where it is an internal error instead. The error is clearly with the import YAML mechanism. What other debugging information do you need that doesn't break my confidentiality?

**Comment by dimitropoulos** [comment]
For the 3 that generate the error, I'm assuming they reliably and repeatably generate the error.  I realize the answer is probably no, but is it possible for you to share these files with us, even in part?  If no, it would be helpful if you could try and narrow down something that can reliably reproduce the error since the above example doesn't cause the error for me anymore.

Any other information you can provide is useful:

- how many lines are the json files that are failing (and how many lines are the ones that aren't failing)
- what operating system are you on
- were you able to run it through a json validator? e.g. https://jsonlint.com/

**Comment by gillescoolen** [comment]
@dimitropoulos https://github.com/MicrosoftDocs/vsts-rest-api-specs/blob/master/specification/build/5.0/build.json as provided by @tobylo.

**Comment by dimitropoulos** [comment]
(the last line of the first paragraph above):
> since the above example doesn't cause the error for me anymore

yep, I was talking about exactly that file.  causes no issue for me.

![Screenshot_20210323_102731](https://user-images.githubusercontent.com/15232461/112162381-62a79c00-8bc2-11eb-9036-f24f975e4eec.png)

**Comment by gillescoolen** [comment]
Weird, maybe it was fixed then, I'll try again this weekend with my old file.

**Comment by gillescoolen** [comment]
So I finally got around to testing, and it seems this issue was fixed!

**Comment by dimitropoulos** [comment]
great! thanks!  if anything along these lines comes up again please don't hesitate to reach out :)

**Comment by daveharding** [comment]
I'm seeing this issue in 2021.3.0 - "Maximum call stack size exceeded" when importing a OpenAPI 3.0 spec file (same issue when trying to import via URL). Filesize is 167KB.

I'm not able to share our API spec file, but is there anything I can try in order to resolve this issue?

![Screenshot 2021-04-30 at 09 37 19](https://user-images.githubusercontent.com/96621/116670590-d09c6b80-a997-11eb-8b4c-40b0daba018d.png)
![Screenshot 2021-04-30 at 09 38 46](https://user-images.githubusercontent.com/96621/116670641-e316a500-a997-11eb-8827-987b44f76708.png)

**Comment by CPlusPlus17** [comment]
Same here, attached file will cause it to crash.
[swagger.json.txt](https://github.com/Kong/insomnia/files/6590757/swagger.json.txt)

**Comment by dimitropoulos** [comment]
thanks @CPlusPlus17 .  I just tried it with the example you provided and got a call stack exceed:

![Screenshot_20210603_135615](https://user-images.githubusercontent.com/15232461/120690634-78172d00-c473-11eb-8c5b-7cb62cf88d31.png)

We'll have to take a look at this.  CC: @reynolek

**Comment by Defkon1** [comment]
same problem here with a swagger json OAS3 file generated from a .NET 5 web api project

**Comment by AGenius** [comment]
I too am now getting this with a expanded api spec - original worked but after adding some new stuff to it it will not import

**Comment by tweetybird99** [comment]
I get the same problem with a openapi specification I downloaded from camunda.    Attached here.   I had to change the extension to .txt

It was originally openapi.json
[openapi.txt](https://github.com/Kong/insomnia/files/6641041/openapi.txt)

Version info here... 
![image](https://user-images.githubusercontent.com/85767484/121750346-34a76900-cada-11eb-86f9-da506d2b119a.png)

**Comment by dimitropoulos** [comment]
thank you for providing another spec @tweetybird99

**Comment by wdawson** [comment]
Sorry folks! This was incorrectly tagged as a feature request along with other bugs. I'm going back through and cleaning up, sorry for the inconvenience.

**Comment by xZyph** [comment]
Get the same issue as the others with a "Maximum call stack size exceeded" error when opening a swagger.json file.

![image](https://user-images.githubusercontent.com/17903817/124352724-7d969d00-dc02-11eb-8cd0-9ad2d2b9e417.png)

**Comment by reynolek** [comment]
@xZyphout of curiosity — how many lines and how many megabytes is the file?

**Comment by xZyph** [comment]
> @xZyphout of curiosity — how many lines and how many megabytes is the file?

@reynolek, I had to check on my phone, but it says 91.94 KB and a total of 3984 lines of JSON data.

**Comment by reynolek** [comment]
Ok so it doesnt seem to be in relation to the size (as I have successfully loaded a 45k+ OAS spec before that was substantially loaded) -- Do you have any kind of reference that could be cyclical inside of it?

**Comment by xZyph** [comment]
There are a few references in there, but when I look through them I don't see anything that looks cyclical. But yeah, a good way to test if this is the problem?

**Comment by xZyph** [comment]
@reynolek, maybe this stack trace will help?

```
RangeError: Maximum call stack size exceeded
    at object (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:397:17)
    at generateParameterExample (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at array (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:410:27)
    at generateParameterExample (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at object (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:402:45)
    at generateParameterExample (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at array (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:410:27)
    at generateParameterExample (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at object (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:402:45)
    at generateParameterExample (C:\ProgramData\user\insomnia\app-2021.4.0\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
````

**Comment by earshinov** [comment]
Same with YouTrack API spec:
https://nv-grt.myjetbrains.com/youtrack/api/openapi.json

**Comment by naefp** [comment]
we have the same problem with our swagger file.

any news about that?

**Comment by xZyph** [comment]
Insomnia looks very promising, but is a "no go" until they fix this issue unfortunately.

**Comment by naefp** [comment]
exactly... I had to change to postman for one of my projects..

**Comment by mmusket** [comment]
I'm also running into the same issue when I try to import a schema 

`RangeError: Maximum call stack size exceeded
    at object (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:397:17)
    at generateParameterExample (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at array (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:410:27)
    at generateParameterExample (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at object (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:402:45)
    at generateParameterExample (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at array (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:410:27)
    at generateParameterExample (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)
    at object (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:402:45)
    at generateParameterExample (...\Local\insomnia\app-2021.5.2\resources\app.asar\node_modules\insomnia-importers\dist\src\importers\openapi-3.js:439:16)`

**Comment by MrSnix** [comment]
Hi folks, I think I may have fixed it (or at least **one of** the root causes of this bug).
I can raise a PR if this sounds reasonable, opinions are welcome, anyway I think it's a step forward.

File used to test:
1. https://nv-grt.myjetbrains.com/youtrack/api/openapi.json (YouTrack API, provided by @earshinov)
2. https://github.com/Kong/insomnia/files/6641041/openapi.txt (Camunda API, rename as .json, provided by @tweetybird99)
3. https://github.com/Kong/insomnia/files/6590757/swagger.json.txt (provided by @CPlusPlus17)

**Before**:
![image](https://user-images.githubusercontent.com/20780192/132408776-2e775d11-2a3c-49d7-950e-c79e133d5980.png)

**After**:
![image](https://user-images.githubusercontent.com/20780192/132408889-a8ef376f-c861-4d35-8398-0f0ebd1fce44.png)
![image](https://user-images.githubusercontent.com/20780192/132411666-0eb65317-f44c-44de-ba9d-e6942cedd567.png)
![image](https://user-images.githubusercontent.com/20780192/132413014-2d0a912d-f626-4c0e-815f-ee0466150373.png)

**Root cause**:
Circular references were killing the stack, I simply turned them off and it started working again.

**How did you fix it?**
```ts
  try {
    apiDocument = await SwaggerParser.validate(apiDocument, {
      dereference: {
        circular: 'ignore', // Here the new option
      },
    }) as OpenAPIV3.Document;
  } catch (err) {
    console.log('[openapi-3] Import file validation failed', err);
  }
```
Please, check the SwaggerParser documentation about this specific option: [Click here](https://github.com/APIDevTools/swagger-parser/blob/main/docs/README.md#circular-refs).

**Thoughts**:
- I think some people with **`really huge (i mean, really really huge)`** file imports may still experience this stack size error but those with a performant system should overcome it.
- One test specs claimed as "affected by the stack size error" was fine for me, I can only assume you didn't have enough power to compute them recursively (see @Xetera answer on the thread about `vsts-rest-api-specs.json`, it was fine for me too.)
- Indeed there was a problem with circular references

**Note**:
- **I found another bug** while I was fixing this one (completely unrelated to this) and I fixed it too.

**Comment by xZyph** [comment]
> **How did you fix it?**
> 
> ```ts
>   try {
>     apiDocument = await SwaggerParser.validate(apiDocument, {
>       dereference: {
>         circular: 'ignore', // Here the new option
>       },
>     }) as OpenAPIV3.Document;
>   } catch (err) {
>     console.log('[openapi-3] Import file validation failed', err);
>   }
> ```

 @MrSnix, is this fix meant to be implemented in Insomnia seeing as the setting is set in the parser, or can I reconfigure my Swagger-instance to somehow exclude the circular references?

**Comment by MrSnix** [comment]
Hi @xZyph

The fix itself is meant to be implemented inside Insomnia but ... you can try to manually modify your specs **excluding (deleting)** all `$ref` fields inside the file. I used this simple regex `"\$ref.*` and it worked but **I cannot guarantee the result,** you will lose the schema references inside your swagger but it's a workaround (so, make a backup). You won't lose them when the fix will be implemented inside Insomnia.

Here is an example of what to do:

**Before**
<img width="400" alt="Example" src="https://user-images.githubusercontent.com/20780192/132490161-f33d8aae-d6d6-4605-a47a-ee23c9c4dd30.png">

**After**
<img width="400" alt="Example" src="https://user-images.githubusercontent.com/20780192/132490301-a613fe4e-e49d-4e6f-ad05-6da0390bd2a0.png">

It worked for me, I hope it will works for you too.

**Comment by naefp** [comment]
i would also appreciate it if insomnia ignores the circular references to fix this issue

**Comment by ac1dr3d** [comment]
great to hear you fixed it

**Comment by karim42benhammou** [comment]
Same issue here. On mac ~20 000 lines OAS 

![Screenshot 2022-02-10 at 18 50 20](https://user-images.githubusercontent.com/91456325/153466724-362e0de6-6a3b-4b43-b443-ba729ca50dbe.png)

**Comment by AmmarHassan1** [comment]
Facing same issue for OAS 3 for 11K lines and  296Kb yaml file. 

The file does has references in it, removing references is like what's the point. 

Version: Insomnia 2021.7.2
Release date: 10/12/2021
OS: Windows_NT x64 10.0.19043
Electron: 11.2.3
Node: 12.18.3
V8: 8.7.220.31-electron.0
Architecture: x64
node-libcurl: libcurl/7.73.0-DEV OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.9 WinIDN libssh2/1.9.0_DEV nghttp2/1.41.0

![image](https://user-images.githubusercontent.com/98140067/154892233-99b157f5-ed52-478d-8fce-2ccaac79325a.png)

**Comment by CPlusPlus17** [comment]
This bug is back after release 2022.06 till the current beta.
